src/model/markerscollection.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import LiveRange from './liverange.js';
+import Range from './range.js';
+import EmitterMixin from '../../utils/emittermixin.js';
+import CKEditorError from '../../utils/ckeditorerror.js';
+import mix from '../../utils/mix.js';
+
+/**
+ * Manages and stores markers.
+ *
+ * Markers are simply {@link module:engine/model/liverange~LiveRange live ranges} added to `MarkersCollection`.
+ * Markers are used to represent information connected with model document. In contrary to
+ * {@link module:engine/model/node~Node nodes}, which are bits of data, markers are marking a part of model document.
+ * Each live range/marker is added with `name` parameter. Name is used to group and identify markers. Multiple live
+ * ranges can be added under the same name.
+ *
+ * Whenever live range is added or removed from markers collection, markers collection fires `addMarker` and `removeMarker`
+ * events. Same happens when a live range is changed due to changes in the model (both events are fired, `removeMarker` first,
+ * then `addMarker`).
+ *
+ * Markers can be converted to view by adding appropriate converters for
+ * {@link module:engine/conversion/modelconversiondispatcher~ModelConversionDispatcher#event:addMarker addMarker} and
+ * {@link module:engine/conversion/modelconversiondispatcher~ModelConversionDispatcher#event:removeMarker removeMarker}
+ * events, or by building converters for {@link module:engine/conversion/modelconversiondispatcher~ModelConversionDispatcher}
+ * using {@link module:engine/conversion/buildmodelconverter~buildModelConverter model converter builder}.
+ *
+ * Markers are similar to adding and converting attributes on nodes. The difference is that attribute is connected to
+ * a given node (i.e. a character is bold no matter if it gets moved or content around it changes). Markers on the
+ * other hand are continuous ranges (i.e. if a character from inside of marker range is moved somewhere else, marker
+ * range is shrunk and the character does not have any attribute or information that it was in the marked range).
+ *
+ * @memberOf engine.model
+ */
+export default class MarkersCollection {
+	/**
+	 * Creates a markers collection.
+	 */
+	constructor() {
+		/**
+		 * Stores range to marker name bindings for added ranges.
+		 *
+		 * @private
+		 * @member {Map} #_markerNames
```

Comment:
Since we'll implement getting, removing and updating markers by name, it makes sense to store `name` => `range` mapping now. I'll rename this property to `_nameToRange`.